### ISSUE TITLE
Concurrent map writes during deferred disconnected context cancellation

### ISSUE THREAD

**Issue by cretz** [temporalio/sdk-go#743]
## Expected Behavior

No concurrent map writes.

## Actual Behavior

This:

```
fatal error: concurrent map writes

goroutine 369355 [running]:
runtime.throw({0x19714cc, 0x28})
	GOROOT/src/runtime/panic.go:1198 +0x71 fp=0xc00295f418 sp=0xc00295f3e8 pc=0x433b31
runtime.mapassign(0xc00297f350, 0xc002ff1800, 0x141b073)
	GOROOT/src/runtime/map.go:585 +0x4d6 fp=0xc00295f498 sp=0xc00295f418 pc=0x40ead6
go.temporal.io/sdk/internal.(*commandsHelper).addCommand(0xc0013b7680, {0x1d2faa0, 0xc00297f350})
	external/io_temporal_go_sdk/internal/internal_decision_state_machine.go:892 +0x1e8 fp=0xc00295f528 sp=0xc00295f498 pc=0x141e328
go.temporal.io/sdk/internal.(*activityCommandStateMachine).cancel(0xc002e6d2d8)
	external/io_temporal_go_sdk/internal/internal_decision_state_machine.go:544 +0x7f fp=0xc00295f558 sp=0xc00295f528 pc=0x141ccdf
go.temporal.io/sdk/internal.(*commandsHelper).requestCancelActivityTask(0xc00295f5b0, {0xc001245310, 0x10000c0005c0800})
	external/io_temporal_go_sdk/internal/internal_decision_state_machine.go:950 +0x39 fp=0xc00295f598 sp=0xc00295f558 pc=0x141eb19
go.temporal.io/sdk/internal.(*workflowEnvironmentImpl).RequestCancelActivity(0xc001c98c00, {{0xc001245310, 0x1}})
	external/io_temporal_go_sdk/internal/internal_event_handlers.go:507 +0x38 fp=0xc00295f608 sp=0xc00295f598 pc=0x1423878
go.temporal.io/sdk/internal.(*workflowExecutionEventHandlerImpl).RequestCancelActivity(0xc001ad0d20, {{0xc001245310, 0x0}})
	<autogenerated>:1 +0x2f fp=0xc00295f630 sp=0xc00295f608 pc=0x146392f
go.temporal.io/sdk/internal.(*workflowEnvironmentInterceptor).ExecuteActivity.func3({0x15bab80, 0xc00000f3c8}, 0x95)
	external/io_temporal_go_sdk/internal/workflow.go:525 +0x75 fp=0xc00295f670 sp=0xc00295f630 pc=0x145cd35
go.temporal.io/sdk/internal.(*channelImpl).Close(0xc000c12a60)
	external/io_temporal_go_sdk/internal/internal_workflow.go:833 +0x93 fp=0xc00295f6c8 sp=0xc00295f670 pc=0x144d653
go.temporal.io/sdk/internal.(*cancelCtx).cancel(0xc000c12a20, 0x1, {0x1cc83a0, 0xc000420ab0})
	external/io_temporal_go_sdk/internal/context.go:333 +0xe9 fp=0xc00295f778 sp=0xc00295f6c8 pc=0x1412b09
go.temporal.io/sdk/internal.NewDisconnectedContext.func1()
	external/io_temporal_go_sdk/internal/context.go:210 +0x30 fp=0xc00295f7a8 sp=0xc00295f778 pc=0x1412230
runtime.deferCallSave(0xc00295f870, 0xc0026ff420)
	GOROOT/src/runtime/panic.go:950 +0x82 fp=0xc00295f7b8 sp=0xc00295f7a8 pc=0x4332c2
runtime.runOpenDeferFrame(0x1000000004056cd, 0xc0023ae550)
	GOROOT/src/runtime/panic.go:889 +0x27b fp=0xc00295f838 sp=0xc00295f7b8 pc=0x432c5b
runtime.Goexit()
	GOROOT/src/runtime/panic.go:642 +0x187 fp=0xc00295f8b8 sp=0xc00295f838 pc=0x432227
go.temporal.io/sdk/internal.(*coroutineState).exit.func1({0xc0029ec000, 0x19ed8a0}, 0xc0009a0701)
	external/io_temporal_go_sdk/internal/internal_workflow.go:938 +0x17 fp=0xc00295f8c8 sp=0xc00295f8b8 pc=0x14624d7
go.temporal.io/sdk/internal.(*coroutineState).initialYield(0xc00162e1e0, 0x3, {0xc0009a0760, 0x1a})
	external/io_temporal_go_sdk/internal/internal_workflow.go:860 +0x89 fp=0xc00295f8f8 sp=0xc00295f8c8 pc=0x144d949
go.temporal.io/sdk/internal.(*coroutineState).yield(...)
	external/io_temporal_go_sdk/internal/internal_workflow.go:869
go.temporal.io/sdk/internal.(*channelImpl).Receive(0xc001dcf4d0, {0x1cf6e50, 0xc000c12a20}, {0x0, 0x0})
	external/io_temporal_go_sdk/internal/internal_workflow.go:684 +0x245 fp=0xc00295f980 sp=0xc00295f8f8 pc=0x144c6e5
go.temporal.io/sdk/internal.(*decodeFutureImpl).Get(0xc002e6d1b8, {0x1cf6e50, 0xc000c12a20}, {0x0, 0x0})
	external/io_temporal_go_sdk/internal/internal_workflow.go:1334 +0x4c fp=0xc00295f9e0 sp=0xc00295f980 pc=0x145122c
[<...workflow-specific-code...>]
go.temporal.io/sdk/internal.(*dispatcherImpl).NewCoroutine.func1(0x0)
	external/io_temporal_go_sdk/internal/internal_workflow.go:971 +0xde fp=0xc00295ffc8 sp=0xc00295ff38 pc=0x144e6be
go.temporal.io/sdk/internal.(*dispatcherImpl).NewCoroutine·dwrap·34()
	external/io_temporal_go_sdk/internal/internal_workflow.go:972 +0x2d fp=0xc00295ffe0 sp=0xc00295ffc8 pc=0x144e5ad
runtime.goexit()
	src/runtime/asm_amd64.s:1581 +0x1 fp=0xc00295ffe8 sp=0xc00295ffe0 pc=0x464101
created by go.temporal.io/sdk/internal.(*dispatcherImpl).NewCoroutine
	external/io_temporal_go_sdk/internal/internal_workflow.go:962 +0x365
```

This seems similar to #721 but a different cancellation path

## Steps to Reproduce the Problem

Still trying to replicate

**Comment by cretz** [comment]
An issue like this has also been fixed in #741. Despite my best efforts, I have not been able to replicate this concurrent map write and it may have been fixed by #721 or #741. Any replication by anyone hitting this in the latest version would be very helpful.

**Comment by thomashusa** [comment]
I may have run into a very similar issue, also during deferred context cancellation:

```


main.main()
'	/src/cmd/worker/main.go:39 +0x229
goroutine 14 [select]:
'	/src/cmd/worker/main.go:83 +0x71
'	/go/pkg/mod/github.com/!data!dog/datadog-go@v4.8.3+incompatible/statsd/sender.go:95 +0xe5
'	/go/pkg/mod/github.com/!data!dog/datadog-go@v4.8.3+incompatible/statsd/sender.go:95 +0xe5
'	<workflow internals>
runtime.deferCallSave(0xc001bf4eb0, 0xc001bf5408)
go.temporal.io/sdk/internal.(*coroutineState).initialYield(0xc003421310, 0x3, {0xc002681480, 0x1b})
'	/go/pkg/mod/go.temporal.io/sdk@v1.14.0/internal/internal_workflow.go:938 +0x17 fp=0xc001bf4f08 sp=0xc001bf4ef8 pc=0xc06f37
'	/usr/local/go/src/runtime/panic.go:661 +0x1ce fp=0xc001bf4ef8 sp=0xc001bf4e78 pc=0x435e4e
go.temporal.io/sdk/internal.(*dispatcherImpl).NewCoroutine·dwrap·34()
'	<workflow internals>
runtime.Goexit()
'	/usr/local/go/src/runtime/panic.go:950 +0x82 fp=0xc001bf4e78 sp=0xc001bf4e68 pc=0x436ea2
'	/go/pkg/mod/go.temporal.io/sdk@v1.14.0/internal/internal_workflow.go:962 +0x365
go.temporal.io/sdk/internal.(*coroutineState).exit.func1({0xc0000a76c0, 0x1a7f2d0}, 0xc002681401)
'	/go/pkg/mod/go.temporal.io/sdk@v1.14.0/internal/internal_workflow.go:971 +0xde fp=0xc001bf5fc8 sp=0xc001bf5f38 pc=0xbf1b1e
created by go.temporal.io/sdk/internal.(*dispatcherImpl).NewCoroutine
'	/go/pkg/mod/go.temporal.io/sdk@v1.14.0/internal/internal_workflow.go:833 +0x93 fp=0xc001bf4be0 sp=0xc001bf4b88 pc=0xbf0ab3
'	/go/pkg/mod/go.temporal.io/sdk@v1.14.0/internal/internal_workflow.go:972 +0x2d fp=0xc001bf5fe0 sp=0xc001bf5fc8 pc=0xbf1a0d
'	/go/pkg/mod/go.temporal.io/sdk@v1.14.0/internal/context.go:196 +0x30 fp=0xc001bf4cc0 sp=0xc001bf4c90 pc=0xbb5050
'	/usr/local/go/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc001bf5fe8 sp=0xc001bf5fe0 pc=0x46aee1
'	<workflow internals>
go.temporal.io/sdk/internal.(*cancelCtx).cancel(0xc003aa4a20, 0x1, {0x1e0ef80, 0xc000216060})
'	/go/pkg/mod/go.temporal.io/sdk@v1.14.0/internal/context.go:333 +0xe9 fp=0xc001bf4c90 sp=0xc001bf4be0 pc=0xbb5aa9
go.temporal.io/sdk/internal.(*channelImpl).Receive(0xc003aa6b40, {0x1e3e078, 0xc000988ea0}, {0x0, 0x0})
go.temporal.io/sdk/internal.(*coroutineState).yield(...)
'	<workflow internals>
'	/go/pkg/mod/go.temporal.io/sdk@v1.14.0/internal/internal_workflow.go:860 +0x89 fp=0xc001bf4f38 sp=0xc001bf4f08 pc=0xbf0da9
go.temporal.io/sdk/internal.(*decodeFutureImpl).Get(0xc00353b200, {0x1e3e078, 0xc000988ea0}, {0x1651fe0, 0xc003ab00d0})
runtime.goexit()
go.temporal.io/sdk/internal.(*dispatcherImpl).NewCoroutine.func1(0x0)
go.temporal.io/sdk/internal.WithCancel.func1()
'	/go/pkg/mod/go.temporal.io/sdk@v1.14.0/internal/internal_workflow.go:1334 +0x4c fp=0xc001bf5020 sp=0xc001bf4fc0 pc=0xbf468c
go.temporal.io/sdk/internal.(*channelImpl).Close(0xc003aa4a60)
'	/go/pkg/mod/go.temporal.io/sdk@v1.14.0/internal/internal_decision_state_machine.go:953 +0x26 fp=0xc001bf4ab0 sp=0xc001bf4a70 pc=0xbc1d66
'	/go/pkg/mod/go.temporal.io/sdk@v1.14.0/internal/workflow.go:526 +0x75 fp=0xc001bf4b88 sp=0xc001bf4b48 pc=0xc01995
go.temporal.io/sdk/internal.(*commandsHelper).requestCancelActivityTask(0xc001bf4b08, {0xc0044d7e00, 0x10000c001bf4af0})
'	/usr/local/go/src/runtime/panic.go:1198 +0x71 fp=0xc001bf49a0 sp=0xc001bf4970 pc=0x437711
'	/usr/local/go/src/runtime/map.go:469 +0x205 fp=0xc001bf49e0 sp=0xc001bf49a0 pc=0x410385
go.temporal.io/sdk/internal.(*workflowExecutionEventHandlerImpl).RequestCancelActivity(0xc000bfc0a8, {{0xc0044d7e00, 0x0}})
go.temporal.io/sdk/internal.(*workflowEnvironmentInterceptor).ExecuteActivity.func3({0x1645ea0, 0xc0047f7f30}, 0xde)
go.temporal.io/sdk/internal.(*workflowEnvironmentImpl).RequestCancelActivity(0xc000530300, {{0xc0044d7e00, 0xc002f5bec0}})
go.temporal.io/sdk/internal.(*commandsHelper).getCommand(0xf, {0x1, {0xc0044d7e00, 0x227447d}})
goroutine 2760 [running]:
runtime.mapaccess2(0x0, 0x0, 0x10)
'	/go/pkg/mod/go.temporal.io/sdk@v1.14.0/internal/internal_decision_state_machine.go:877 +0x65 fp=0xc001bf4a70 sp=0xc001bf49e0 pc=0xbc1285
runtime.throw({0x1a087d3, 0x0})
'	<autogenerated>:1 +0x2f fp=0xc001bf4b48 sp=0xc001bf4b20 pc=0xc083ef
'	/go/pkg/mod/go.temporal.io/sdk@v1.14.0/internal/internal_event_handlers.go:507 +0x38 fp=0xc001bf4b20 sp=0xc001bf4ab0 pc=0xbc6ad8
fatal error: concurrent map read and map write
```

**Comment by cretz** [comment]
@thomashusa - Is it possible to replicate this in anyway? Using the `-race` flag could make it more likely to happen. This can intentionally happen if you're using Go async constructs inside your workflow, but if not, we definitely want to replicate and solev this. As of yet, we've been unable to replicate.

**Comment by thomashusa** [comment]
@cretz So far we haven't been able to reliably reproduce it. For us it happens only in production under a lot of load. We're hesitant to run in production with the `-race` flag enabled and are still trying to see if we can reproduce it in development.

**Comment by cretz** [comment]
https://github.com/temporalio/sdk-go/blob/269b55b8d184849209c6973fe913a427fafc07e3/internal/context.go#L239 seems to start a Goroutine in workflow context which is likely the cause. Will require some effort to fix and confirm compatible.

**Comment by tdeebswihart** [comment]
[This may have just happened in CI](https://github.com/temporalio/sdk-go/actions/runs/6227934624/job/16903653066):

```
2023-09-18T21:09:05.0968374Z WARNING: DATA RACE
2023-09-18T21:09:05.0968666Z Read at 0x00c000250fc0 by goroutine 6063:
2023-09-18T21:09:05.0968983Z   runtime.mapaccess2()
2023-09-18T21:09:05.0969463Z       /opt/hostedtoolcache/go/1.20.8/x64/src/runtime/map.go:456 +0x0
2023-09-18T21:09:05.0969904Z   go.temporal.io/sdk/internal.(*commandsHelper).addCommand()
2023-09-18T21:09:05.0970610Z       /home/runner/work/sdk-go/sdk-go/internal/internal_command_state_machine.go:929 +0xb6
2023-09-18T21:09:05.0971095Z   go.temporal.io/sdk/internal.(*commandsHelper).recordVersionMarker()
2023-09-18T21:09:05.0971828Z       /home/runner/work/sdk-go/sdk-go/internal/internal_command_state_machine.go:1073 +0x512
2023-09-18T21:09:05.0972322Z   go.temporal.io/sdk/internal.(*workflowEnvironmentImpl).GetVersion()
2023-09-18T21:09:05.0973013Z       /home/runner/work/sdk-go/sdk-go/internal/internal_event_handlers.go:799 +0x734
2023-09-18T21:09:05.0973539Z   go.temporal.io/sdk/internal.(*workflowExecutionEventHandlerImpl).GetVersion()
2023-09-18T21:09:05.0973955Z       <autogenerated>:1 +0x78
2023-09-18T21:09:05.0974389Z   go.temporal.io/sdk/internal.(*workflowEnvironmentInterceptor).GetVersion()
2023-09-18T21:09:05.0975037Z       /home/runner/work/sdk-go/sdk-go/internal/workflow.go:1681 +0x6f
2023-09-18T21:09:05.0975529Z   go.temporal.io/sdk/internal.(*WorkflowOutboundInterceptorBase).GetVersion()
2023-09-18T21:09:05.0976364Z       /home/runner/work/sdk-go/sdk-go/internal/interceptor_base.go:336 +0x9a
2023-09-18T21:09:05.0976885Z   go.temporal.io/sdk/test_test.(*tracingWorkflowOutboundInterceptor).GetVersion()
2023-09-18T21:09:05.0977304Z       <autogenerated>:1 +0x29
2023-09-18T21:09:05.0977640Z   go.temporal.io/sdk/internal.GetVersion()
2023-09-18T21:09:05.0978227Z       /home/runner/work/sdk-go/sdk-go/internal/workflow.go:1677 +0xb7
2023-09-18T21:09:05.0978618Z   go.temporal.io/sdk/workflow.GetVersion()
2023-09-18T21:09:05.0979192Z       /home/runner/work/sdk-go/sdk-go/workflow/workflow.go:384 +0x19e
2023-09-18T21:09:05.0979638Z   go.temporal.io/sdk/test_test.(*Workflows).VersionLoopWorkflow()
2023-09-18T21:09:05.0980251Z       /home/runner/work/sdk-go/sdk-go/test/workflow_test.go:2319 +0x113
2023-09-18T21:09:05.0980789Z   go.temporal.io/sdk/test_test.(*Workflows).VersionLoopWorkflow-fm()
2023-09-18T21:09:05.0981183Z       <autogenerated>:1 +0x92
2023-09-18T21:09:05.0981460Z   runtime.call64()
2023-09-18T21:09:05.0982008Z       /opt/hostedtoolcache/go/1.20.8/x64/src/runtime/asm_amd64.s:730 +0x48
2023-09-18T21:09:05.0983328Z   reflect.Value.Call()
2023-09-18T21:09:05.0983837Z       /opt/hostedtoolcache/go/1.20.8/x64/src/reflect/value.go:370 +0xc7
2023-09-18T21:09:05.0984248Z   go.temporal.io/sdk/internal.executeFunction()
2023-09-18T21:09:05.0984905Z       /home/runner/work/sdk-go/sdk-go/internal/internal_worker.go:1806 +0x5b0
2023-09-18T21:09:05.0985426Z   go.temporal.io/sdk/internal.(*workflowEnvironmentInterceptor).ExecuteWorkflow()
2023-09-18T21:09:05.0986090Z       /home/runner/work/sdk-go/sdk-go/internal/workflow.go:530 +0x23e
2023-09-18T21:09:05.0986595Z   go.temporal.io/sdk/internal.(*WorkflowInboundInterceptorBase).ExecuteWorkflow()
2023-09-18T21:09:05.0987285Z       /home/runner/work/sdk-go/sdk-go/internal/interceptor_base.go:154 +0x76
2023-09-18T21:09:05.0987812Z   go.temporal.io/sdk/test_test.(*signalWorkflowInboundInterceptor).ExecuteWorkflow()
2023-09-18T21:09:05.0988260Z       <autogenerated>:1 +0x29
2023-09-18T21:09:05.0988742Z   go.temporal.io/sdk/test_test.(*tracingWorkflowInboundInterceptor).ExecuteWorkflow()
2023-09-18T21:09:05.0989427Z       /home/runner/work/sdk-go/sdk-go/test/integration_test.go:4136 +0x18e
2023-09-18T21:09:05.0989848Z   go.temporal.io/sdk/internal.(*workflowExecutor).Execute()
2023-09-18T21:09:05.0990477Z       /home/runner/work/sdk-go/sdk-go/internal/internal_worker.go:805 +0x41a
2023-09-18T21:09:05.0990958Z   go.temporal.io/sdk/internal.(*syncWorkflowDefinition).Execute.func1()
2023-09-18T21:09:05.0991621Z       /home/runner/work/sdk-go/sdk-go/internal/internal_workflow.go:512 +0x189
2023-09-18T21:09:05.0992092Z   go.temporal.io/sdk/internal.(*dispatcherImpl).NewCoroutine.func1()
2023-09-18T21:09:05.0992760Z       /home/runner/work/sdk-go/sdk-go/internal/internal_workflow.go:1057 +0x165
2023-09-18T21:09:05.0995837Z   go.temporal.io/sdk/internal.(*dispatcherImpl).NewCoroutine.func2()
2023-09-18T21:09:05.0996566Z       /home/runner/work/sdk-go/sdk-go/internal/internal_workflow.go:1058 +0x47
2023-09-18T21:09:05.0996819Z 
2023-09-18T21:09:05.0996995Z Previous write at 0x00c000250fc0 by goroutine 6062:
2023-09-18T21:09:05.0997322Z   runtime.mapdelete()
2023-09-18T21:09:05.0997796Z       /opt/hostedtoolcache/go/1.20.8/x64/src/runtime/map.go:695 +0x0
2023-09-18T21:09:05.0998255Z   go.temporal.io/sdk/internal.(*commandStateMachineBase).moveState()
2023-09-18T21:09:05.0998968Z       /home/runner/work/sdk-go/sdk-go/internal/internal_command_state_machine.go:430 +0x646
2023-09-18T21:09:05.0999482Z   go.temporal.io/sdk/internal.(*markerCommandStateMachine).handleCommandSent()
2023-09-18T21:09:05.1000210Z       /home/runner/work/sdk-go/sdk-go/internal/internal_command_state_machine.go:836 +0xab
2023-09-18T21:09:05.1000675Z   go.temporal.io/sdk/internal.(*commandsHelper).getCommands()
2023-09-18T21:09:05.1001372Z       /home/runner/work/sdk-go/sdk-go/internal/internal_command_state_machine.go:1417 +0x276
2023-09-18T21:09:05.1001919Z   go.temporal.io/sdk/internal.(*workflowExecutionContextImpl).CompleteWorkflowTask()
2023-09-18T21:09:05.1002929Z       /home/runner/work/sdk-go/sdk-go/internal/internal_task_handlers.go:1244 +0x81b
2023-09-18T21:09:05.1003469Z   go.temporal.io/sdk/internal.(*workflowExecutionContextImpl).applyWorkflowPanicPolicy()
2023-09-18T21:09:05.1004227Z       /home/runner/work/sdk-go/sdk-go/internal/internal_task_handlers.go:1142 +0x1064
2023-09-18T21:09:05.1004764Z   go.temporal.io/sdk/internal.(*workflowExecutionContextImpl).ProcessWorkflowTask()
2023-09-18T21:09:05.1005492Z       /home/runner/work/sdk-go/sdk-go/internal/internal_task_handlers.go:1088 +0x1676
2023-09-18T21:09:05.1005934Z   go.temporal.io/sdk/internal.executeDispatcher()
2023-09-18T21:09:05.1006564Z       /home/runner/work/sdk-go/sdk-go/internal/internal_workflow.go:622 +0xc7
2023-09-18T21:09:05.1007053Z   go.temporal.io/sdk/internal.(*syncWorkflowDefinition).OnWorkflowTaskStarted()
2023-09-18T21:09:05.1007734Z       /home/runner/work/sdk-go/sdk-go/internal/internal_workflow.go:592 +0x88
2023-09-18T21:09:05.1008785Z   go.temporal.io/sdk/internal.(*workflowExecutionEventHandlerImpl).ProcessEvent()
2023-09-18T21:09:05.1009528Z       /home/runner/work/sdk-go/sdk-go/internal/internal_event_handlers.go:1063 +0x602
2023-09-18T21:09:05.1010062Z   go.temporal.io/sdk/internal.(*workflowExecutionContextImpl).ProcessWorkflowTask()
2023-09-18T21:09:05.1010803Z       /home/runner/work/sdk-go/sdk-go/internal/internal_task_handlers.go:1028 +0x1fc9
2023-09-18T21:09:05.1011317Z   go.temporal.io/sdk/internal.(*workflowTaskHandlerImpl).ProcessWorkflowTask()
2023-09-18T21:09:05.1012007Z       /home/runner/work/sdk-go/sdk-go/internal/internal_task_handlers.go:836 +0x69b
2023-09-18T21:09:05.1012511Z   go.temporal.io/sdk/internal.(*workflowTaskPoller).processWorkflowTask()
2023-09-18T21:09:05.1013204Z       /home/runner/work/sdk-go/sdk-go/internal/internal_task_pollers.go:354 +0x755
2023-09-18T21:09:05.1013676Z   go.temporal.io/sdk/internal.(*workflowTaskPoller).ProcessTask()
2023-09-18T21:09:05.1014339Z       /home/runner/work/sdk-go/sdk-go/internal/internal_task_pollers.go:318 +0xca
2023-09-18T21:09:05.1014790Z   go.temporal.io/sdk/internal.(*baseWorker).processTask()
2023-09-18T21:09:05.1015419Z       /home/runner/work/sdk-go/sdk-go/internal/internal_worker_base.go:505 +0x255
2023-09-18T21:09:05.1015896Z   go.temporal.io/sdk/internal.(*baseWorker).processTaskAsync.func1()
2023-09-18T21:09:05.1016566Z       /home/runner/work/sdk-go/sdk-go/internal/internal_worker_base.go:356 +0x8b
2023-09-18T21:09:05.1016801Z 
2023-09-18T21:09:05.1016941Z Goroutine 6063 (running) created at:
2023-09-18T21:09:05.1017346Z   go.temporal.io/sdk/internal.(*dispatcherImpl).NewCoroutine()
2023-09-18T21:09:05.1017994Z       /home/runner/work/sdk-go/sdk-go/internal/internal_workflow.go:1048 +0x66e
2023-09-18T21:09:05.1018466Z   go.temporal.io/sdk/internal.(*workflowEnvironmentInterceptor).Go()
2023-09-18T21:09:05.1019140Z       /home/runner/work/sdk-go/sdk-go/internal/internal_workflow.go:308 +0x86
2023-09-18T21:09:05.1019646Z   go.temporal.io/sdk/test_test.(*tracingWorkflowOutboundInterceptor).Go()
2023-09-18T21:09:05.1020338Z       /home/runner/work/sdk-go/sdk-go/test/integration_test.go:4093 +0x201
2023-09-18T21:09:05.1020750Z   go.temporal.io/sdk/internal.newDispatcher()
2023-09-18T21:09:05.1021373Z       /home/runner/work/sdk-go/sdk-go/internal/internal_workflow.go:614 +0x27e
2023-09-18T21:09:05.1021823Z   go.temporal.io/sdk/internal.(*syncWorkflowDefinition).Execute()
2023-09-18T21:09:05.1022467Z       /home/runner/work/sdk-go/sdk-go/internal/internal_workflow.go:499 +0x1c4
2023-09-18T21:09:05.1023048Z   go.temporal.io/sdk/internal.(*workflowExecutionEventHandlerImpl).handleWorkflowExecutionStarted()
2023-09-18T21:09:05.1023824Z       /home/runner/work/sdk-go/sdk-go/internal/internal_event_handlers.go:1269 +0x36b
2023-09-18T21:09:05.1024353Z   go.temporal.io/sdk/internal.(*workflowExecutionEventHandlerImpl).ProcessEvent()
2023-09-18T21:09:05.1025074Z       /home/runner/work/sdk-go/sdk-go/internal/internal_event_handlers.go:1044 +0x20e
2023-09-18T21:09:05.1025755Z   go.temporal.io/sdk/internal.(*workflowExecutionContextImpl).ProcessWorkflowTask()
2023-09-18T21:09:05.1026473Z       /home/runner/work/sdk-go/sdk-go/internal/internal_task_handlers.go:1028 +0x1fc9
2023-09-18T21:09:05.1026989Z   go.temporal.io/sdk/internal.(*workflowTaskHandlerImpl).ProcessWorkflowTask()
2023-09-18T21:09:05.1027695Z       /home/runner/work/sdk-go/sdk-go/internal/internal_task_handlers.go:836 +0x69b
2023-09-18T21:09:05.1028191Z   go.temporal.io/sdk/internal.(*workflowTaskPoller).processWorkflowTask()
2023-09-18T21:09:05.1028882Z       /home/runner/work/sdk-go/sdk-go/internal/internal_task_pollers.go:354 +0x755
2023-09-18T21:09:05.1029351Z   go.temporal.io/sdk/internal.(*workflowTaskPoller).ProcessTask()
2023-09-18T21:09:05.1030000Z       /home/runner/work/sdk-go/sdk-go/internal/internal_task_pollers.go:318 +0xca
2023-09-18T21:09:05.1030504Z   go.temporal.io/sdk/internal.(*baseWorker).processTask()
2023-09-18T21:09:05.1031151Z       /home/runner/work/sdk-go/sdk-go/internal/internal_worker_base.go:505 +0x255
2023-09-18T21:09:05.1031811Z   go.temporal.io/sdk/internal.(*baseWorker).processTaskAsync.func1()
2023-09-18T21:09:05.1032493Z       /home/runner/work/sdk-go/sdk-go/internal/internal_worker_base.go:356 +0x8b
2023-09-18T21:09:05.1032729Z 
2023-09-18T21:09:05.1032873Z Goroutine 6062 (running) created at:
2023-09-18T21:09:05.1033263Z   go.temporal.io/sdk/internal.(*baseWorker).processTaskAsync()
2023-09-18T21:09:05.1034094Z       /home/runner/work/sdk-go/sdk-go/internal/internal_worker_base.go:352 +0x139
2023-09-18T21:09:05.1034559Z   go.temporal.io/sdk/internal.(*baseWorker).runTaskDispatcher()
2023-09-18T21:09:05.1035221Z       /home/runner/work/sdk-go/sdk-go/internal/internal_worker_base.go:382 +0x116
2023-09-18T21:09:05.1035651Z   go.temporal.io/sdk/internal.(*baseWorker).Start.func3()
2023-09-18T21:09:05.1036302Z       /home/runner/work/sdk-go/sdk-go/internal/internal_worker_base.go:284 +0x39
```

**Comment by Quinn-With-Two-Ns** [comment]
@tdeebswihart That failure is unrelated to this ticket, both are data races, but the root cause it different. It should be in it's own separate ticket.

**Comment by Quinn-With-Two-Ns** [comment]
For anyone reading this issue, the code @cretz pointed was actually dead code and not the cause of the problem. The root cause was actually a race due to how the SDK evicts workflows from cache, more details [here](https://github.com/temporalio/sdk-go/issues/1209). 

There is still some follow up work, but the bug described in this ticket should be resolved in the latest Go SDK `1.25.1`
* https://github.com/temporalio/sdk-go/issues/1270
* https://github.com/temporalio/sdk-go/issues/1235
